### PR TITLE
Use a separate animation ID from the session ID in all cases, including initial animation.

### PR DESCRIPTION
**Pull request by megangardner**
#### c5c0cb2d749aea4fd57661a4bbdfd4bfca617dbc
<pre>
Use a separate animation ID from the session ID in all cases, including initial animation.
<a href="https://bugs.webkit.org/show_bug.cgi?id=275700">https://bugs.webkit.org/show_bug.cgi?id=275700</a>
<a href="https://rdar.apple.com/130217887">rdar://130217887</a>

Reviewed by Richard Robinson.

Perviously, I was using the sessionUUID for the first animation. This has made it difficult
to write clean and compartmentalized code, and caused issues where the wrong UUID was used.
This patch separated these into two separate concepts, and added mapping storage in the animation
controller. I also clean up some of the naming for the clean up code, and generally make the code
easier to understand.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::removeTextAnimationForAnimationID):
(WebCore::ChromeClient::removeTransparentMarkersForSessionID):
(WebCore::ChromeClient::removeInitialTextAnimation):
(WebCore::ChromeClient::addInitialTextAnimation):
(WebCore::ChromeClient::removeTextAnimationForID): Deleted.
(WebCore::ChromeClient::cleanUpTextAnimationsForSessionID): Deleted.
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::didBeginWritingToolsSession):
(WebCore::WritingToolsController::proofreadingSessionDidReceiveSuggestions):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
(WebCore::WritingToolsController::didEndWritingToolsSession):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextAnimationForAnimationID:withData:]):
(-[WKWebView _removeTextAnimationForAnimationID:]):
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView _enableSourceTextAnimationAfterElementWithID:]):
(-[WKWebView _enableFinalTextAnimationForElementWithID:]):
(-[WKWebView _disableTextAnimationWithUUID:]):
(-[WKWebView _addTextAnimationTypeForID:withData:]): Deleted.
(-[WKWebView _removeTextAnimationForID:]): Deleted.
(-[WKWebView beginWritingToolsAnimationForSessionWithUUID:]): Deleted.
(-[WKWebView endWritingToolsAnimationForSessionWithUUID:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::addTextAnimationForAnimationID):
(WebKit::PageClientImplCocoa::removeTextAnimationForAnimationID):
(WebKit::PageClientImplCocoa::addTextAnimationTypeForID): Deleted.
(WebKit::PageClientImplCocoa::removeTextAnimationForID): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextAnimationForAnimationID):
(WebKit::WebPageProxy::removeTextAnimationForAnimationID):
(WebKit::WebPageProxy::addTextAnimationTypeForID): Deleted.
(WebKit::WebPageProxy::removeTextAnimationForID): Deleted.
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WKSTextAnimationManager.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView addTextAnimationForAnimationID:withStyleType:]):
(-[WKContentView removeTextAnimationForAnimationID:]):
(-[WKContentView addTextAnimationTypeForID:withStyleType:]): Deleted.
(-[WKContentView removeTextAnimationForID:]): Deleted.
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.h:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
(-[WKTextAnimationManager removeTextAnimationForAnimationID:]):
(-[WKTextAnimationManager restoreTextAnimationType]):
(-[WKTextAnimationManager addTextAnimationTypeForID:withData:]): Deleted.
(-[WKTextAnimationManager removeTextAnimationForID:]): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::addTextAnimationForAnimationID):
(WebKit::WebViewImpl::removeTextAnimationForAnimationID):
(WebKit::WebViewImpl::addTextAnimationTypeForID): Deleted.
(WebKit::WebViewImpl::removeTextAnimationForID): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
(beginEffect(for:style:)):
(endEffect(for:)):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::removeTextAnimationForAnimationID):
(WebKit::WebChromeClient::removeTransparentMarkersForSessionID):
(WebKit::WebChromeClient::removeInitialTextAnimation):
(WebKit::WebChromeClient::addInitialTextAnimation):
(WebKit::WebChromeClient::removeTextAnimationForID): Deleted.
(WebKit::WebChromeClient::cleanUpTextAnimationsForSessionID): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h: Renamed from Source/WebKit/WebProcess/WebPage/TextAnimationController.h.
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
(WebKit::TextAnimationController::contextRangeForTextAnimationID const):
(WebKit::TextAnimationController::removeTransparentMarkersForSessionID):
(WebKit::TextAnimationController::removeInitialTextAnimation):
(WebKit::TextAnimationController::addInitialTextAnimation):
(WebKit::TextAnimationController::addSourceTextAnimation):
(WebKit::TextAnimationController::addDestinationTextAnimation):
(WebKit::TextAnimationController::updateUnderlyingTextVisibilityForTextAnimationID):
(WebKit::TextAnimationController::createTextIndicatorForTextAnimationID):
(WebKit::TextAnimationController::contextRangeForTextAnimationType const): Deleted.
(WebKit::TextAnimationController::cleanUpTextAnimationsForSessionID): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextAnimationForAnimationID):
(WebKit::WebPage::removeTextAnimationForAnimationID):
(WebKit::WebPage::removeTransparentMarkersForSessionID):
(WebKit::WebPage::removeInitialTextAnimation):
(WebKit::WebPage::addInitialTextAnimation):
(WebKit::WebPage::addTextAnimationTypeForID): Deleted.
(WebKit::WebPage::removeTextAnimationForID): Deleted.
(WebKit::WebPage::cleanUpTextAnimationsForSessionID): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/280319@main">https://commits.webkit.org/280319@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0f4399f5f4da2e2f2347a90975a012561e1f1aa9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56267 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8739 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/59873 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6703 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58393 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43215 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45554 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4661 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58296 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33477 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48544 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26431 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30256 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5874 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5707 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52264 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6146 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61556 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/175 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6273 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52834 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/175 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48610 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52720 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12464 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/no-raf-while-render-blocked.html (failure)") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/156 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8353 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31420 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32506 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33589 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32253 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/ChromeClient.h [hunk 1]
```diff
@@ -668,9 +668,13 @@ class ChromeClient {
...
 
-    virtual void removeTextAnimationForID(const WritingTools::SessionID&) { }
+    virtual void removeTextAnimationForAnimationID(const WritingTools::SessionID&) { }
 
-    virtual void cleanUpTextAnimationsForSessionID(const WritingTools::SessionID&) { }
+    virtual void removeTransparentMarkersForSessionID(const WritingTools::SessionID&) { }
+
+    virtual void removeInitialTextAnimation(const WritingTools::SessionID&) { }
+
+    virtual void addInitialTextAnimation(const WritingTools::SessionID&) { }
```
Comment (rr-codes) [inline Source/WebCore/page/ChromeClient.h]: maybe these should be named
```
removeTextAnimationForAnimationWithID
removeTransparentMarkersForSessionWithID
removeInitialTextAnimationForSessionWithID
addInitialTextAnimationForSessionWithID
```

also, for the first one, it probably shouldn't take a `SessionID` type

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm [hunk 2]
```diff
@@ -93,17 +93,23 @@
     return corePage->contextRangeForSessionWithID(sessionID);
 }
 
-std::optional<WebCore::SimpleRange> TextAnimationController::contextRangeForTextAnimationType(const WTF::UUID& uuid) const
+std::optional<WebCore::SimpleRange> TextAnimationController::contextRangeForTextAnimationID(const WTF::UUID& animationUUID) const
 {
-    if (auto sessionRange = contextRangeForSessionWithID(uuid))
+    // FIXME: This shouldn't be needed any longer, but still is.
```
Comment (rr-codes) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: >but still is

why? this seems like it can't ever return a non-nil value if you're using an animation id for a session id?

File: Source/WebCore/page/ChromeClient.h (hunk 1 above)
Comment (megangardner) [reply Source/WebCore/page/ChromeClient.h]: I don't know that I agree with the 'with', seems extra wordy for no reason.

Comment (rr-codes) [reply Source/WebCore/page/ChromeClient.h]: ok; I think the latter two methods should at least then have `ForSessionID` added to them to avoid confusion (and the parameter type of the first one should not be SessionID)

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm [hunk 3]
```diff
@@ -93,17 +93,23 @@
...
         return sessionRange;
 
-    if (m_manuallyEnabledAnimationRanges.contains(uuid))
-        return WebCore::makeSimpleRange(m_manuallyEnabledAnimationRanges.find(uuid)->value.get());
+    if (m_manuallyEnabledAnimationRanges.contains(animationUUID))
+        return WebCore::makeSimpleRange(m_manuallyEnabledAnimationRanges.find(animationUUID)->value.get());
+
+    for (auto sessionUUID : m_initialAnimations.keys()) {
+        if (m_initialAnimations.get(sessionUUID) == animationUUID)
+            return contextRangeForSessionWithID(sessionUUID);
```
Comment (rr-codes) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:103]: should this map instead be keyed by animation id and have a session id value, since in this case you're trying to find a key based on a value?

Comment (megangardner) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:103]: I want them all keyed the same way for now, once things settle it might make sense to optimize them in different ways.

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm [hunk 4]
```diff
@@ -138,6 +146,12 @@
     });
 }
 
+void TextAnimationController::removeInitialTextAnimation(const WTF::UUID& sessionUUID)
+{
+    if (m_initialAnimations.contains(sessionUUID))
+        m_webPage->removeTextAnimationForAnimationID(m_initialAnimations.get(sessionUUID));
```
Comment (rr-codes) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: I think this is slightly more efficient / cleaner if it's done like

```
if (CheckedPtr animationID = m_initialAnimations.find(sessionUUID))
    m_webPage->removeTextAnimationForAnimationID(*animationID);
```

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm [hunk 5]
```diff
@@ -112,17 +118,19 @@
...
     if (animationStates.isEmpty())
         return;
     for (auto animationState : animationStates)
         removeTransparentMarkersForTextAnimationID(animationState.styleID);
 
-    auto unstyledRange = m_unstyledRanges.find(sessionUUID);
-    if (unstyledRange->value)
-        removeTransparentMarkersForTextAnimationID(unstyledRange->value->styleID);
+    if (m_unstyledRanges.contains(sessionUUID))
+        removeTransparentMarkersForTextAnimationID(m_unstyledRanges.find(sessionUUID)->value->styleID);
```
Comment (rr-codes) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: ditto for this; also, what happens if `value` is `nullptr`?

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm [hunk 6]
```diff
@@ -93,17 +93,23 @@
...
 {
-    if (auto sessionRange = contextRangeForSessionWithID(uuid))
+    // FIXME: This shouldn't be needed any longer, but still is.
+    if (auto sessionRange = contextRangeForSessionWithID(animationUUID))
         return sessionRange;
 
-    if (m_manuallyEnabledAnimationRanges.contains(uuid))
-        return WebCore::makeSimpleRange(m_manuallyEnabledAnimationRanges.find(uuid)->value.get());
+    if (m_manuallyEnabledAnimationRanges.contains(animationUUID))
+        return WebCore::makeSimpleRange(m_manuallyEnabledAnimationRanges.find(animationUUID)->value.get());
```
Comment (rr-codes) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: ditto about using `find` in the `if` statement

**Comment by pxlcoder** [comment]
Typo in commit title: `seperate` -> `separate`

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm (hunk 5 above)
Comment (megangardner) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: That's what the contains check is for, or do I not understand how our HashMaps are implemented?

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm (hunk 2 above)
Comment (megangardner) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: Oh this is leftover from when I was trying to figure out the crashes, this should be removed,  I think.

File: Source/WebCore/page/writing-tools/WritingToolsController.mm [hunk 7]
```diff
@@ -182,6 +182,10 @@
 void WritingToolsController::didBeginWritingToolsSession(const WritingTools::Session& session, const Vector<WritingTools::Context>& contexts)
 {
     RELEASE_LOG(WritingTools, "WritingToolsController::didBeginWritingToolsSession (%s) [received contexts: %zu]", session.identifier.toString().utf8().data(), contexts.size());
+
+    // Don't animate smart replies, they are animated by UIKit/AppKit.
+    if (session.compositionType != WebCore::WritingTools::Session::CompositionType::SmartReply)
+        m_page->chrome().client().addInitialTextAnimation(session.identifier);
```
Comment (rr-codes) [inline Source/WebCore/page/writing-tools/WritingToolsController.mm:209]: why is this added?

Comment (megangardner) [reply Source/WebCore/page/writing-tools/WritingToolsController.mm:209]: It's moved, this is a better place to add this from, I don't have to pipe down to the WebProcess separately.

File: Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift [hunk 8]
```diff
@@ -41,7 +41,7 @@ import WebKitSwift
         delegate.containingViewForTextAnimationType().addSubview(self.effectView)
     }
     
-    @objc(addTextAnimationTypeForID:withStyleType:) public func beginEffect(for uuid: UUID, style: WKTextAnimationType) {
+    @objc(addTextAnimationTypeForAnimationID:withStyleType:) public func beginEffect(for uuid: UUID, style: WKTextAnimationType) {
```
Comment (rr-codes) [inline Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift]: I think this should be `addTextAnimationForAnimationID` since you're adding the animation and not just what type it is

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm [hunk 9]
```diff
@@ -112,17 +114,19 @@
...
     if (animationStates.isEmpty())
         return;
     for (auto animationState : animationStates)
         removeTransparentMarkersForTextAnimationID(animationState.styleID);
 
-    auto unstyledRange = m_unstyledRanges.find(sessionUUID);
-    if (unstyledRange->value)
-        removeTransparentMarkersForTextAnimationID(unstyledRange->value->styleID);
+    if (auto iterator = m_unstyledRanges.find(sessionUUID); iterator != m_unstyledRanges.end())
+        removeTransparentMarkersForTextAnimationID(iterator->value->styleID);
```
Comment (rr-codes) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: if `value` is `nullptr` here, it can cause a crash, right?

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h [hunk 10]
```diff
@@ -84,12 +86,14 @@ class TextAnimationController final {
...
 
 private:
-    std::optional<WebCore::SimpleRange> contextRangeForTextAnimationType(const WTF::UUID&) const;
+    std::optional<WebCore::SimpleRange> contextRangeForTextAnimationID(const WTF::UUID&) const;
     std::optional<WebCore::SimpleRange> contextRangeForSessionWithID(const WebCore::WritingTools::SessionID&) const;
 
     RefPtr<WebCore::Document> document() const;
     WeakPtr<WebPage> m_webPage;
 
+    // All of these HashMap keys are SessionUUIDs
```
Comment (rr-codes) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h]: `SessionID` (not `SessionUUID`)

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm (hunk 9 above)
Comment (megangardner) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: How would it be null? We know we have one.

**Review by rr-codes — APPROVED** [review]
